MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/TextureRender.java

```diff
@@ -0,0 +1,243 @@
+            GLES20.glDeleteProgram(program);
+            Log.w(TAG, "Failed to link program");
+            return 0;
+        }
+
+        return program;
+    }
+
+    private int compileShader(int type, String shaderSrc) {
+        Log.d(TAG, "Compiling shader: " + type);
+
+        // Create new shader object
+        final int shader = GLES20.glCreateShader(type);
+        if (shader == 0) {
+            Log.w(TAG, "Could not create new shader");
+            return 0;
+        }
+
+        // Pass in the shader source
+        GLES20.glShaderSource(shader, shaderSrc);
+
+        // Compile the shader
+        GLES20.glCompileShader(shader);
+
+        // Verify the compile status
+        final int[] compiled = new int[1];
+        GLES20.glGetShaderiv(shader, GLES20.GL_COMPILE_STATUS, compiled, /* offset = */ 0);
+        if (compiled[0] == 0) {
+            // Delete the shader if compilation failed
+            GLES20.glDeleteShader(shader);
+            Log.w(TAG, "Shader failed to compile");
+            return 0;
+        }
+
+        return shader;
+    }
+
+    public void drawFrame(SurfaceTexture surfaceTexture) {
+        Log.d(TAG, "Drawing frame");
+
+        // Set the transform matrix of the surface texture
+        surfaceTexture.getTransformMatrix(uMatrix);
+
+        GLES20.glClearColor(/* red = */ 0.0f, /* green = */ 0.0f, /* blue = */ 0.0f, /* alpha = */ 1.0f);
+        GLES20.glClear(/* mask = */ GLES20.GL_COLOR_BUFFER_BIT);
+
+        GLES20.glUseProgram(program);
+
+        // Process vertices
+        triangleVertices.position(TRIANGLE_VERTICES_DATA_POS_OFFSET);
+        GLES20.glVertexAttribPointer(
+                aPositionHandle,
+                /* size = */ 3,
+                /* type = */ GLES20.GL_FLOAT,
+                /* normalized = */ false,
+                TRIANGLE_VERTICES_DATA_STRIDE_BYTES,
+                triangleVertices
+        );
+        GLES20.glEnableVertexAttribArray(aPositionHandle);
+
+        // Process texture
+        triangleVertices.position(TRIANGLE_VERTICES_DATA_UV_OFFSET);
+        GLES20.glVertexAttribPointer(
+                aTextureHandle,
+                /* size = */ 2,
+                /* type = */ GLES20.GL_FLOAT,
+                /* normalized = */ false,
+                TRIANGLE_VERTICES_DATA_STRIDE_BYTES,
+                triangleVertices
+        );
+        GLES20.glEnableVertexAttribArray(aTextureHandle);
+
+        // Apply matrix transforms
+        GLES20.glUniformMatrix4fv(
+                uMatrixHandle,
+                /* count = */ 1,
+                /* transpose = */ false,
+                uMatrix,
+                /* offset = */ 0
+        );
+
+        GLES20.glActiveTexture(GLES20.GL_TEXTURE0);
+        GLES20.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, textureID);
+        GLES20.glUniform1i(uTextureUnitHandle, /* x = */0);
+
+        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, /* first = */ 0, /* count = */ 4);
+        GLES20.glFinish();
```

Comment:
Is there a reason why you use glFinish here? This is generally not needed / not a good practice.

Reply:
Changes: deleted